File: src/resolution/name-matcher.ts
/**
 * Name Matcher
 *
 * Handles symbol name matching for reference resolution.
 */

import { Node } from '../types';
import { UnresolvedRef, ResolvedRef, ResolutionContext } from './types';

/**
 * Try to resolve a reference by exact name match
 */
export function matchByExactName(
  ref: UnresolvedRef,
  context: ResolutionContext
): ResolvedRef | null {
  const candidates = context.getNodesByName(ref.referenceName);

  if (candidates.length === 0) {
    return null;
  }

  // If only one match, use it
  if (candidates.length === 1) {
    return {
      original: ref,
      targetNodeId: candidates[0]!.id,
      confidence: 0.9,
      resolvedBy: 'exact-match',
    };
  }

  // Multiple matches - try to narrow down
  const bestMatch = findBestMatch(ref, candidates, context);
  if (bestMatch) {
    // Lower confidence when the match is from a distant/unrelated module
    const proximity = computePathProximity(ref.filePath, bestMatch.filePath);
    const confidence = proximity >= 30 ? 0.7 : 0.4;
    return {
      original: ref,
      targetNodeId: bestMatch.id,
      confidence,
      resolvedBy: 'exact-match',
    };
  }

  return null;
}

/**
 * Try to resolve by qualified name
 */
export function matchByQualifiedName(
  ref: UnresolvedRef,
  context: ResolutionContext
): ResolvedRef | null {
  // Check if the reference name looks qualified (contains :: or .)
  if (!ref.referenceName.includes('::') && !ref.referenceName.includes('.')) {
    return null;
  }

  const candidates = context.getNodesByQualifiedName(ref.referenceName);

  if (candidates.length === 1) {
    return {
      original: ref,
      targetNodeId: candidates[0]!.id,
      confidence: 0.95,
      resolvedBy: 'qualified-name',
    };
  }

  // Try partial qualified name match
  const parts = ref.referenceName.split(/[:.]/);
  const lastName = parts[parts.length - 1];
  if (lastName) {
    const partialCandidates = context.getNodesByName(lastName);
    for (const candidate of partialCandidates) {
      if (candidate.qualifiedName.endsWith(ref.referenceName)) {
        return {
          original: ref,
          targetNodeId: candidate.id,
          confidence: 0.85,
          resolvedBy: 'qualified-name',
        };
      }
    }
  }

  return null;
}

/**
 * Try to resolve by method name on a class/object
 */
export function matchMethodCall(
  ref: UnresolvedRef,
  context: ResolutionContext
): ResolvedRef | null {
  // Parse method call patterns like "obj.method" or "Class::method"
  const dotMatch = ref.referenceName.match(/^(\w+)\.(\w+)$/);
  const colonMatch = ref.referenceName.match(/^(\w+)::(\w+)$/);

  const match = dotMatch || colonMatch;
  if (!match) {
    return null;
  }

  const [, objectOrClass, methodName] = match;

  // Find the class/object first
  const classCandidates = context.getNodesByName(objectOrClass!);

  for (const classNode of classCandidates) {
    if (classNode.kind === 'class' || classNode.kind === 'struct' || classNode.kind === 'interface') {
      // Look for method in the same file
      const nodesInFile = context.getNodesInFile(classNode.filePath);
      const methodNode = nodesInFile.find(
        (n) =>
          n.kind === 'method' &&
          n.name === methodName &&
          n.qualifiedName.includes(classNode.name)
      );

      if (methodNode) {
        return {
          original: ref,
          targetNodeId: methodNode.id,
          confidence: 0.85,
          resolvedBy: 'qualified-name',
        };
      }
    }
  }

  return null;
}

/**
 * Compute directory proximity between two file paths.
 * Returns a score based on the number of shared directory segments.
 * Higher score = closer in directory tree.
 */
function computePathProximity(filePath1: string, filePath2: string): number {
  const dir1 = filePath1.split('/').slice(0, -1);
  const dir2 = filePath2.split('/').slice(0, -1);

  let shared = 0;
  for (let i = 0; i < Math.min(dir1.length, dir2.length); i++) {
    if (dir1[i] === dir2[i]) {
      shared++;
    } else {
      break;
    }
  }

  // Each shared directory segment contributes 15 points, capped at 80
  return Math.min(shared * 15, 80);
}

/**
 * Find the best matching node when there are multiple candidates
 */
function findBestMatch(
  ref: UnresolvedRef,
  candidates: Node[],
  _context: ResolutionContext
): Node | null {
  // Prioritization rules:
  // 1. Same file > different file
  // 2. Directory proximity (same module/package > different module)
  // 3. Same language > different language
  // 4. Functions/methods > classes/types (for call references)
  // 5. Exported > non-exported

  let bestScore = -1;
  let bestNode: Node | null = null;

  for (const candidate of candidates) {
    let score = 0;

    // Same file bonus
    if (candidate.filePath === ref.filePath) {
      score += 100;
    }

    // Directory proximity bonus — strongly prefer same module/package
    score += computePathProximity(ref.filePath, candidate.filePath);

    // Same language bonus
    if (candidate.language === ref.language) {
      score += 50;
    }

    // For call references, prefer functions/methods
    if (ref.referenceKind === 'calls') {
      if (candidate.kind === 'function' || candidate.kind === 'method') {
        score += 25;
      }
    }

    // Exported bonus
    if (candidate.isExported) {
      score += 10;
    }

    // Closer line number (within same file)
    if (candidate.filePath === ref.filePath && candidate.startLine) {
      const distance = Math.abs(candidate.startLine - ref.line);
      score += Math.max(0, 20 - distance / 10);
    }

    if (score > bestScore) {
      bestScore = score;
      bestNode = candidate;
    }
  }

  return bestNode;
}

/**
 * Fuzzy match - last resort with lower confidence
 */
export function matchFuzzy(
  ref: UnresolvedRef,
  context: ResolutionContext
): ResolvedRef | null {
  const lowerName = ref.referenceName.toLowerCase();

  // Use pre-built lowercase index for O(1) lookup instead of scanning all nodes
  const candidates = context.getNodesByLowerName(lowerName);

  // Filter to callable kinds only (function, method, class)
  const callableKinds = new Set(['function', 'method', 'class']);
  const callableCandidates = candidates.filter((n) => callableKinds.has(n.kind));

  if (callableCandidates.length === 1) {
    return {
      original: ref,
      targetNodeId: callableCandidates[0]!.id,
      confidence: 0.5,
      resolvedBy: 'fuzzy',
    };
  }

  return null;
}

/**
 * Match all strategies in order of confidence
 */
export function matchReference(
  ref: UnresolvedRef,
  context: ResolutionContext
): ResolvedRef | null {
  // Try strategies in order of confidence
  let result: ResolvedRef | null;

  // 1. Qualified name match (highest confidence)
  result = matchByQualifiedName(ref, context);
  if (result) return result;

  // 2. Method call pattern
  result = matchMethodCall(ref, context);
  if (result) return result;

  // 3. Exact name match
  result = matchByExactName(ref, context);
  if (result) return result;

  // 4. Fuzzy match (lowest confidence)
  result = matchFuzzy(ref, context);
  if (result) return result;

  return null;
}

File: go/internal/resolution/name_matcher.go
package resolution

import (
	"strings"

	"github.com/kristofer/codegraph/internal/types"
)

// MatchReference tries all name-matching strategies in order of confidence and
// returns the best result, or nil if nothing could be matched.
func MatchReference(ref *types.UnresolvedReference, ctx *ResolutionContext) *ResolvedRef {
	if result := matchByFilePath(ref, ctx); result != nil {
		return result
	}
	if result := matchByQualifiedName(ref, ctx); result != nil {
		return result
	}
	if result := matchMethodCall(ref, ctx); result != nil {
		return result
	}
	if result := matchByExactName(ref, ctx); result != nil {
		return result
	}
	return matchFuzzy(ref, ctx)
}

// matchByFilePath resolves path-like references (e.g. "snippets/drawer.liquid")
// by matching the filename segment against file nodes.
func matchByFilePath(ref *types.UnresolvedReference, ctx *ResolutionContext) *ResolvedRef {
	if !strings.Contains(ref.ReferenceName, "/") {
		return nil
	}
	parts := strings.Split(ref.ReferenceName, "/")
	fileName := parts[len(parts)-1]
	if fileName == "" {
		return nil
	}

	candidates := ctx.GetNodesByName(fileName)
	var fileNodes []*types.Node
	for _, n := range candidates {
		if n.Kind == types.NodeKindFile {
			fileNodes = append(fileNodes, n)
		}
	}
	if len(fileNodes) == 0 {
		return nil
	}

	// Prefer exact path match
	for _, n := range fileNodes {
		if n.QualifiedName == ref.ReferenceName || n.FilePath == ref.ReferenceName {
			return &ResolvedRef{
				FromNodeID: ref.FromNodeID, TargetNodeID: n.ID,
				Confidence: 0.95, ResolvedBy: "file-path",
			}
		}
	}
	// Suffix match
	for _, n := range fileNodes {
		if strings.HasSuffix(n.QualifiedName, ref.ReferenceName) || strings.HasSuffix(n.FilePath, ref.ReferenceName) {
			return &ResolvedRef{
				FromNodeID: ref.FromNodeID, TargetNodeID: n.ID,
				Confidence: 0.85, ResolvedBy: "file-path",
			}
		}
	}
	if len(fileNodes) == 1 {
		return &ResolvedRef{
			FromNodeID: ref.FromNodeID, TargetNodeID: fileNodes[0].ID,
			Confidence: 0.7, ResolvedBy: "file-path",
		}
	}
	return nil
}

// matchByQualifiedName resolves references that contain "::" or "." separators
// by looking up the qualified name directly.
func matchByQualifiedName(ref *types.UnresolvedReference, ctx *ResolutionContext) *ResolvedRef {
	if !strings.Contains(ref.ReferenceName, "::") && !strings.Contains(ref.ReferenceName, ".") {
		return nil
	}

	candidates := ctx.GetNodesByQualifiedName(ref.ReferenceName)
	if len(candidates) == 1 {
		return &ResolvedRef{
			FromNodeID: ref.FromNodeID, TargetNodeID: candidates[0].ID,
			Confidence: 0.95, ResolvedBy: "qualified-name",
		}
	}

	// Partial match: try the last segment
	parts := strings.FieldsFunc(ref.ReferenceName, func(r rune) bool {
		return r == ':' || r == '.'
	})
	lastName := parts[len(parts)-1]
	if lastName != "" {
		for _, n := range ctx.GetNodesByName(lastName) {
			if strings.HasSuffix(n.QualifiedName, ref.ReferenceName) {
				return &ResolvedRef{
					FromNodeID: ref.FromNodeID, TargetNodeID: n.ID,
					Confidence: 0.85, ResolvedBy: "qualified-name",
				}
			}
		}
	}
	return nil
}

// matchMethodCall handles "obj.method" and "Class::method" patterns.
func matchMethodCall(ref *types.UnresolvedReference, ctx *ResolutionContext) *ResolvedRef {
	var objectOrClass, methodName string

	if idx := strings.Index(ref.ReferenceName, "::"); idx > 0 {
		objectOrClass = ref.ReferenceName[:idx]
		methodName = ref.ReferenceName[idx+2:]
	} else if idx := strings.Index(ref.ReferenceName, "."); idx > 0 {
		objectOrClass = ref.ReferenceName[:idx]
		methodName = ref.ReferenceName[idx+1:]
	}

	if objectOrClass == "" || methodName == "" {
		return nil
	}

	// Strategy 1: direct class name match
	for _, classNode := range ctx.GetNodesByName(objectOrClass) {
		if classNode.Kind != types.NodeKindClass && classNode.Kind != types.NodeKindStruct && classNode.Kind != types.NodeKindInterface {
			continue
		}
		if classNode.Language != ref.Language {
			continue
		}
		for _, n := range ctx.GetNodesInFile(classNode.FilePath) {
			if n.Kind == types.NodeKindMethod && n.Name == methodName &&
				strings.Contains(n.QualifiedName, classNode.Name) {
				return &ResolvedRef{
					FromNodeID: ref.FromNodeID, TargetNodeID: n.ID,
					Confidence: 0.85, ResolvedBy: "qualified-name",
				}
			}
		}
	}

	// Strategy 2: capitalised receiver (e.g., permissionEngine → PermissionEngine)
	capitalized := strings.ToUpper(objectOrClass[:1]) + objectOrClass[1:]
	if capitalized != objectOrClass {
		for _, classNode := range ctx.GetNodesByName(capitalized) {
			if classNode.Kind != types.NodeKindClass && classNode.Kind != types.NodeKindStruct && classNode.Kind != types.NodeKindInterface {
				continue
			}
			if classNode.Language != ref.Language {
				continue
			}
			for _, n := range ctx.GetNodesInFile(classNode.FilePath) {
				if n.Kind == types.NodeKindMethod && n.Name == methodName &&
					strings.Contains(n.QualifiedName, classNode.Name) {
					return &ResolvedRef{
						FromNodeID: ref.FromNodeID, TargetNodeID: n.ID,
						Confidence: 0.8, ResolvedBy: "instance-method",
					}
				}
			}
		}
	}

	// Strategy 3: scan all methods with this name; score by word overlap with receiver
	if methodName != "" {
		methodCandidates := ctx.GetNodesByName(methodName)
		var methods []*types.Node
		for _, n := range methodCandidates {
			if n.Kind == types.NodeKindMethod {
				methods = append(methods, n)
			}
		}
		// Prefer same language
		var sameLang []*types.Node
		for _, n := range methods {
			if n.Language == ref.Language {
				sameLang = append(sameLang, n)
			}
		}
		target := methods
		if len(sameLang) > 0 {
			target = sameLang
		}
		if len(target) == 1 && target[0].Language == ref.Language {
			return &ResolvedRef{
				FromNodeID: ref.FromNodeID, TargetNodeID: target[0].ID,
				Confidence: 0.7, ResolvedBy: "instance-method",
			}
		}
		if len(target) > 1 {
			receiverWords := splitCamelCase(objectOrClass)
			var best *types.Node
			bestScore := 0
			for _, m := range target {
				classWords := splitCamelCase(m.QualifiedName)
				score := 0
				for _, rw := range receiverWords {
					for _, cw := range classWords {
						if strings.EqualFold(cw, rw) {
							score++
						}
					}
				}
				if m.Language == ref.Language {
					score++
				}
				if score > bestScore {
					bestScore = score
					best = m
				}
			}
			if best != nil && bestScore >= 2 {
				return &ResolvedRef{
					FromNodeID: ref.FromNodeID, TargetNodeID: best.ID,
					Confidence: 0.65, ResolvedBy: "instance-method",
				}
			}
		}
	}
	return nil
}

// matchByExactName resolves by exact simple name match.
func matchByExactName(ref *types.UnresolvedReference, ctx *ResolutionContext) *ResolvedRef {
	candidates := ctx.GetNodesByName(ref.ReferenceName)
	if len(candidates) == 0 {
		return nil
	}
	if len(candidates) == 1 {
		crossLang := candidates[0].Language != ref.Language
		conf := 0.9
		if crossLang {
			conf = 0.5
		}
		return &ResolvedRef{
			FromNodeID: ref.FromNodeID, TargetNodeID: candidates[0].ID,
			Confidence: conf, ResolvedBy: "exact-match",
		}
	}
	best := findBestMatch(ref, candidates)
	if best == nil {
		return nil
	}
	prox := computePathProximity(ref.FilePath, best.FilePath)
	conf := 0.7
	if prox < 30 {
		conf = 0.4
	}
	return &ResolvedRef{
		FromNodeID: ref.FromNodeID, TargetNodeID: best.ID,
		Confidence: conf, ResolvedBy: "exact-match",
	}
}

// matchFuzzy resolves by case-insensitive name match.
func matchFuzzy(ref *types.UnresolvedReference, ctx *ResolutionContext) *ResolvedRef {
	lowerName := strings.ToLower(ref.ReferenceName)
	candidates := ctx.GetNodesByLowerName(lowerName)

	callableKinds := map[types.NodeKind]bool{
		types.NodeKindFunction: true,
		types.NodeKindMethod:   true,
		types.NodeKindClass:    true,
	}
	var callable []*types.Node
	for _, n := range candidates {
		if callableKinds[n.Kind] {
			callable = append(callable, n)
		}
	}

	var sameLanguage []*types.Node
	for _, n := range callable {
		if n.Language == ref.Language {
			sameLanguage = append(sameLanguage, n)
		}
	}
	final := callable
	if len(sameLanguage) > 0 {
		final = sameLanguage
	}

	if len(final) == 1 {
		crossLang := final[0].Language != ref.Language
		conf := 0.5
		if crossLang {
			conf = 0.3
		}
		return &ResolvedRef{
			FromNodeID: ref.FromNodeID, TargetNodeID: final[0].ID,
			Confidence: conf, ResolvedBy: "fuzzy",
		}
	}
	return nil
}

// ===========================================================================
// Helpers
// ===========================================================================

// splitCamelCase splits a camelCase or PascalCase identifier into words.
func splitCamelCase(s string) []string {
	// Insert a space before each uppercase letter that follows a lowercase letter
	var b strings.Builder
	runes := []rune(s)
	for i, r := range runes {
		if i > 0 && r >= 'A' && r <= 'Z' {
			prev := runes[i-1]
			if prev >= 'a' && prev <= 'z' {
				b.WriteRune(' ')
			}
		}
		b.WriteRune(r)
	}
	raw := b.String()
	var words []string
	for _, w := range strings.FieldsFunc(raw, func(r rune) bool {
		return r == ' ' || r == '_' || r == '.' || r == ':' || r == '/' || r == '\\'
	}) {
		if len(w) > 1 {
			words = append(words, w)
		}
	}
	return words
}

// computePathProximity returns a score based on shared directory segments
// (higher = closer in the directory tree), capped at 80.
func computePathProximity(path1, path2 string) int {
	dir1 := strings.Split(path1, "/")
	dir1 = dir1[:len(dir1)-1]
	dir2 := strings.Split(path2, "/")
	dir2 = dir2[:len(dir2)-1]

	shared := 0
	for i := 0; i < len(dir1) && i < len(dir2); i++ {
		if dir1[i] == dir2[i] {
			shared++
		} else {
			break
		}
	}
	score := shared * 15
	if score > 80 {
		score = 80
	}
	return score
}

// findBestMatch scores multiple candidates and returns the best one.
func findBestMatch(ref *types.UnresolvedReference, candidates []*types.Node) *types.Node {
	bestScore := -1000
	var best *types.Node
	for _, c := range candidates {
		score := 0
		if c.FilePath == ref.FilePath {
			score += 100
		}
		score += computePathProximity(ref.FilePath, c.FilePath)
		if c.Language == ref.Language {
			score += 50
		} else {
			score -= 80
		}
		if ref.ReferenceKind == types.EdgeKindCalls {
			if c.Kind == types.NodeKindFunction || c.Kind == types.NodeKindMethod {
				score += 25
			}
		}
		if c.IsExported {
			score += 10
		}
		if c.FilePath == ref.FilePath && c.StartLine > 0 {
			dist := ref.Line - c.StartLine
			if dist < 0 {
				dist = -dist
			}
			add := 20 - dist/10
			if add < 0 {
				add = 0
			}
			score += add
		}
		if score > bestScore {
			bestScore = score
			best = c
		}
	}
	return best
}
